keps/sig-node/5224-node-resource-discovery/README.md

```diff
@@ -0,0 +1,1211 @@
+number of logical CPU cores from the Kubernetes. I want to do this to optimize
+application performance by taking advantage of the detailed knowledge of how
+the hardware works behind the scenes.
+
+
+#### Story 3
+
+As a system administrator, I want my cluster to be aware of the compute
+available throughout its lifespan and improve scheduling based on the node’s
+compute capabilities.
+
+#### Story 4
+
+As a system administrator, I want the cluster to gain insights of the host’s
+architecture specific capabilities and leverage the same for improved
+performance.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Reported capacity does not fit currently running pods. This is more of a
+concern in of [KEP-3953][kep-3953]. In practice, the kubelet/Kubernetes cannot
+prevent resources changing (e.g. CPUs getting offlined).
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### CRI API
+
+A new streaming RPC is added to get the node resources.
+
+The proposed API for exposing resources takes inspiration from
+[NodeResourceTopology API][noderesourcetopology-api]. It allows
+representing the resources in a tree structure representing the hardware
+topology of the system.
+
+```proto
+import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+
+// GetDynamicRuntimeConfig is a streaming interface for receiving dynamically
+// changing runtime and node configuration
+rpc GetDynamicRuntimeConfig(DynamicRuntimeConfigRequest) returns (stream DynamicRuntimeConfigResponse) {}
+
+message DynamicRuntimeConfigRequest{}
+
+message DynamicRuntimeConfigResponse {
+    ResourceTopology resource_topology = 1;
+}
+
+message ResourceTopology {
+    repeated ResourceTopologyZone zones = 1;
+}
+
+// ResourceTopologyZone represents a node in the topology tree
+message ResourceTopologyZone {
+    string name = 1;
+    string type = 2;
+    string parent = 3;
+    repeated ResourceTopologyCost costs = 4;
+    map<string, string> attributes = 5;
+    repeated ResourceTopologyResourceInfo resources = 6;
+}
+
+message ResourceTopologyCost {
+    string name = 1;
+    uint32 value = 2;
+}
+
+message ResourceTopologyResourceInfo {
+    string name = 1;
+    k8s.io.apimachinery.pkg.api.resource.Quantity capacity = 2;
+}
+```
+
+The well-known zone types and attributes that kubelet understands are defined
+as consts.
+
+```go
+const (
+    ResourceTopologyZoneCore       = "Core"
```

Comment:
these currently are a subset of the cadvisor machine info fields. is this all the kubelet uses?

Reply:
Yup, picked the ones that the kubelet is interested in. Of course we can add all that we can imagine (e.g. the current cpu topology levels from linux kernel, i.e. package, die, cluster, core)